README.md
Twitter-Trends
==============

UC Berkeley CS61A


CHANGES
VERSION 1.1
CHANGE:: cache::
    """
    # Storing original implementations of ADTs
    trends.tweet_adt = (make_tweet, tweet_text, tweet_time, tweet_location)
    trends.tweet_fn_adt = (make_tweet_fn, tweet_text_fn, tweet_time_fn, tweet_location_fn)
    trends.sentiment_adt = (make_sentiment, has_sentiment, sentiment_value)
    trends.position_adt = (make_position, latitude, longitude)
    round5 = lambda args: tuple(round(arg, 5) for arg in args)
    trends.database_adt = (make_database, add_value, get_keys, get_values, get_value_from_key, get_len, get_items)
    geo.database_adt = (make_database, add_value, get_items)
    geo.us_states_adt = us_states

    # These classes test abstraction violations
    class AbstractionViolation(Exception):
        pass

    def datatype(obj):
        return type(obj).__name__

    # Generic abstract data type
    class Abstract(object):
        def __add__(self, other):
            raise AbstractionViolation("Can't add {} object to {}".format(datatype(self), datatype(other)))

        def __radd__(self, other):
            raise AbstractionViolation("Can't add {} object to {}".format(datatype(self), datatype(other)))

        def __eq__(self, other):
            if isinstance(other, type(self)):
                return other is self
            raise AbstractionViolation("Can't use == on {} object and {}".format(datatype(self), datatype(other)))

        def __ne__(self, other):
            if isinstance(other, type(self)):
                return other is not self
            raise AbstractionViolation("Can't use != on {} object and {}".format(datatype(self), datatype(other)))

        def __bool__(self):
            raise AbstractionViolation("Can't use {} object as a boolean".format(datatype(self)))

        def __getitem__(self, index):
            raise AbstractionViolation("Can't use [] notation on {} object".format(datatype(self)))

        def __contains__(self, other):
            raise AbstractionViolation("Can't use contains notation on {} object".format(datatype(self)))

        def __delitem__(self, other):
            raise AbstractionViolation("Can't use del notation on {} object".format(datatype(self)))

        def __iter__(self):
            raise AbstractionViolation("Can't iterate on {} object".format(datatype(self)))

        def __len__(self):
            raise AbstractionViolation("Can't use len notation on {} object".format(datatype(self)))

        def __setitem__(self, key, item):
            raise AbstractionViolation("Can't use setitem notation on {} object".format(datatype(self)))


    class Database(Abstract):
        def __init__(self):
            self._dictionary = {}

        @staticmethod
        def make_database():
            return Database()

        def add_value(self, key, value):
            if type(key) != int and type(key) != str:
                raise TypeError("databases only accept strings or numbers as keys")
            self._dictionary[key] = value
            return self

        def get_keys(self):
            return list(self._dictionary.keys())

        def get_values(self):
            return list(self._dictionary.values())

        def get_value_from_key(self, key):
            return self._dictionary[key]

        def get_len(self):
            return len(self._dictionary)

        def get_items(self):
            items = []
            for key, value in self._dictionary.items():
                items += [[key, value]]
            return items

        def get(self, key, default=None):
            raise AbstractionViolation("Can't use get notation on {} object".format(datatype(self)))

        def setdefault(self, key, default=None):
            raise AbstractionViolation("Can't use setdefault notation on {} object".format(datatype(self)))

        def items(self):
            raise AbstractionViolation("Can't use items notation on {} object".format(datatype(self)))

        def keys(self):
            raise AbstractionViolation("Can't use keys notation on {} object".format(datatype(self)))

        def values(self):
            raise AbstractionViolation("Can't use values notation on {} object".format(datatype(self)))

    class Tweet(Abstract):
        def __init__(self, text, time, lat, lon):
            self._text = text
            self._time = time
            self._lat = lat
            self._lon = lon

        def text(self):
            return self._text

        def time(self):
            return self._time

        def location(self):
            return trends.make_position(self._lat, self._lon)

    class Sentiment(Abstract):
        def __init__(self, value):
            assert value is None or type(value) in (float, int), 'invalid value to sentiment constructor: {}'.format(datatype(value))
            self._value = value

        def sentiment_value(self):
            if type(self) != Sentiment:
                raise AbstractionViolation("Can't call sentiment_value on {}".format(self))
            assert self.has_sentiment(), 'No sentiment value'
            return self._value

        def has_sentiment(self):
            if type(self) != Sentiment:
                raise AbstractionViolation("Can't call has_sentiment on {}".format(self))
            return self._value != None

    class Position(Abstract):
        def __init__(self, lat, lon):
            self._lat = lat
            self._lon = lon

        def latitude(self):
            if type(self) != Position:
                raise AbstractionViolation("Can't call latitude on {}".format(self))
            return self._lat

        def longitude(self):
            if type(self) != Position:
                raise AbstractionViolation("Can't call longitude on {}".format(self))
            return self._lon
    """


geo.py
"""Geography and projection utilities."""

from data import DATA_PATH
from math import sin, cos, atan2, radians, sqrt
from json import JSONDecoder
from database import make_database, add_value, get_items

def make_position(lat, lon):
    """Return a geographic position, which has a latitude and longitude."""
    return (lat, lon)

def latitude(position):
    """Return the latitudinal coordinate of a geographic position."""
    return position[0]

def longitude(position):
    """Return the longitudinal coordinate of a geographic position."""
    return position[1]

def geo_distance(position1, position2):
    """Return the great circle distance (in miles) between two
    geographic positions.

    Uses the "haversine" formula.
    http://en.wikipedia.org/wiki/Haversine_formula

    >>> round(geo_distance(make_position(50, 5), make_position(58, 3)), 1)
    559.2
    """
    earth_radius = 3963.2  # miles
    lat1, lat2 = [radians(latitude(p)) for p in (position1, position2)]
    lon1, lon2 = [radians(longitude(p)) for p in (position1, position2)]
    dlat, dlon = lat2-lat1, lon2-lon1
    a = sin(dlat/2) ** 2  + sin(dlon/2) ** 2 * cos(lat1) * cos(lat2)
    c = 2 * atan2(sqrt(a), sqrt(1-a));
    return earth_radius * c;

def position_to_xy(position):
    """Convert a geographic position within the US to a planar x-y point."""
    lat = latitude(position)
    lon = longitude(position)
    if lat < 25:
        return _hawaii(position)
    elif lat > 52:
        return _alaska(position)
    else:
        return _lower48(position)

def albers_projection(origin, parallels, translate, scale):
    """Return an Albers projection from geographic positions to x-y positions.

    Derived from Mike Bostock's Albers javascript implementation for D3
    http://mbostock.github.com/d3
    http://mathworld.wolfram.com/AlbersEqual-AreaConicProjection.html

    origin -- a geographic position
    parallels -- bounding latitudes
    translate -- x-y translation to place the projection within a larger map
    scale -- scaling factor
    """
    phi1, phi2 = [radians(p) for p in parallels]
    base_lat = radians(latitude(origin))
    s, c = sin(phi1), cos(phi1)
    base_lon = radians(longitude(origin))
    n = 0.5 * (s + sin(phi2))
    C = c*c + 2*n*s
    p0 = sqrt(C - 2*n*sin(base_lat))/n

    def project(position):
        lat, lon = radians(latitude(position)), radians(longitude(position))
        t = n * (lon - base_lon)
        p = sqrt(C - 2*n*sin(lat))/n
        x = scale * p * sin(t) + translate[0]
        y = scale * (p * cos(t) - p0) + translate[1]
        return (x, y)
    return project

_lower48 = albers_projection(make_position(38, -98), [29.5, 45.5], [480,250], 1000)
_alaska = albers_projection(make_position(60, -160), [55,65], [150,440], 400)
_hawaii = albers_projection(make_position(20, -160), [8,18], [300,450], 1000)

def load_states():
    """Load the coordinates of all the state outlines and return them
    in a database, from names to shapes lists.

    >>> len(get_value_from_key(load_states(), 'HI'))  # Hawaii has 5 islands
    5
    """
    json_data_file = open(DATA_PATH + 'states.json', encoding='utf8')
    states = JSONDecoder().decode(json_data_file.read())
    states_database = make_database()
    for state, shapes in states.items():
        states_database = add_value(states_database, state, shapes)
    for state, shapes in get_items(states_database):
        for index, shape in enumerate(shapes):
            if type(shape[0][0]) == list:  # the shape is a single polygon
                assert len(shape) == 1, 'Multi-polygon shape'
                shape = shape[0]
            shapes[index] = [make_position(*reversed(pos)) for pos in shape]
    return states_database

us_states = load_states()


database.py
def make_database():
    """ 
    Creates a database
    """
    return []

def add_value(database, key, value):
    """
    Creates a mapping between that key and value inside of the database. If the
    key already exists in the database, then we replace the previous key-value pair
    with the new one.
    """
    if type(key) != int and type(key) != str:
        raise TypeError("databases only accept strings or numbers as keys")
    for i in range(len(database)):
        k, v = database[i]
        if k == key:
            return database[:i] + database[i+1:] + [[key, value]]
    return database + [[key, value]]

def get_keys(database):
    """
    Returns a list that contains all the keys found in that database, in
    arbitrary order
    """
    return [key for key, val in database]

def get_values(database):
    """
    Returns a list that contains all the values found in that database, in
    arbitrary order
    """
    return [val for key, val in database]

def get_value_from_key(database, key):
    """
    Returns the value associated with that key. It will raise error if the key
    is not found in the database.
    """
    for k, v in database:
        if k == key:
            return v
    raise KeyError("key", key, "not found in database")

def get_len(database):
    """
    Returns how many key-value pairs are in the database
    """
    return len(database)

def get_items(database):
    """
    Returns a list of the key-value pairs, in arbitrary order
    """
    return database


trends.py
"""Visualizing Twitter Sentiment Across America"""

from database import make_database, add_value, get_keys, get_values, get_value_from_key, get_items, get_len
from data import word_sentiments, load_tweets
from datetime import datetime
from geo import us_states, geo_distance, make_position, longitude, latitude
from maps import draw_state, draw_name, draw_dot, wait
from string import ascii_letters
from ucb import main, trace, interact, log_current_line


###################################
# Phase 1: The Feelings in Tweets #
###################################

# The tweet abstract data type, implemented as a list.

def make_tweet(text, time, lat, lon):
    """Return a tweet, represented as a Python list.

    text  -- A string; the text of the tweet, all in lowercase
    time  -- A datetime object; the time that the tweet was posted
    lat   -- A number; the latitude of the tweet's location
    lon   -- A number; the longitude of the tweet's location

    >>> t = make_tweet("just ate lunch", datetime(2012, 9, 24, 13), 38, 74)
    >>> tweet_text(t)
    'just ate lunch'
    >>> tweet_time(t)
    datetime.datetime(2012, 9, 24, 13, 0)
    >>> p = tweet_location(t)
    >>> latitude(p)
    38
    >>> tweet_string(t)
    '"just ate lunch" @ (38, 74)'
    """
    return [text, time, lat, lon]

def tweet_text(tweet):
    """Return a string, the words in the text of a tweet."""
    return tweet[0]

def tweet_time(tweet):
    """Return the datetime representing when a tweet was posted."""
    "*** YOUR CODE HERE ***"
    return tweet[1]

def tweet_location(tweet):
    """Return a position representing a tweet's location."""
    "*** YOUR CODE HERE ***"
    return make_position(tweet[2], tweet[3])

# The tweet abstract data type, implemented as a function.

def make_tweet_fn(text, time, lat, lon):
    """An alternate implementation of make_tweet: a tweet is a function.

    >>> t = make_tweet_fn("just ate lunch", datetime(2012, 9, 24, 13), 38, 74)
    >>> tweet_text_fn(t)
    'just ate lunch'
    >>> tweet_time_fn(t)
    datetime.datetime(2012, 9, 24, 13, 0)
    >>> latitude(tweet_location_fn(t))
    38
    """
    "*** YOUR CODE HERE ***"
    # Please don't call make_tweet in your solution
    def make_tweet_fn_key_return(key):
        assert key == 'text' or key == 'time' or key == 'lat' or key == 'lon'
        if key == 'text':
            return text
        elif key == 'time':
            return time
        elif key == 'lat':
            return lat
        elif key == 'lon':
            return lon
        else:
            pass
    return make_tweet_fn_key_return

def tweet_text_fn(tweet):
    """Return a string, the words in the text of a functional tweet."""
    return tweet('text')

def tweet_time_fn(tweet):
    """Return the datetime representing when a functional tweet was posted."""
    return tweet('time')

def tweet_location_fn(tweet):
    """Return a position representing a functional tweet's location."""
    return make_position(tweet('lat'), tweet('lon'))

### === +++ ABSTRACTION BARRIER +++ === ###

def tweet_words(tweet):
    """Return the words in a tweet."""
    return extract_words(tweet_text(tweet))

def tweet_string(tweet):
    """Return a string representing a functional tweet."""
    location = tweet_location(tweet)
    point = (latitude(location), longitude(location))
    return '"{0}" @ {1}'.format(tweet_text(tweet), point)

def extract_words(text):
    """Return the words in a tweet, not including punctuation.

    >>> extract_words('anything else.....not my job')
    ['anything', 'else', 'not', 'my', 'job']
    >>> extract_words('i love my job. #winning')
    ['i', 'love', 'my', 'job', 'winning']
    >>> extract_words('make justin # 1 by tweeting #vma #justinbieber :)')
    ['make', 'justin', 'by', 'tweeting', 'vma', 'justinbieber']
    >>> extract_words("paperclips! they're so awesome, cool, & useful!")
    ['paperclips', 'they', 're', 'so', 'awesome', 'cool', 'useful']
    >>> extract_words('@(cat$.on^#$my&@keyboard***@#*')
    ['cat', 'on', 'my', 'keyboard']
    """
    "*** YOUR CODE HERE ***"
    for i in range(len(text)):
        if text[i] not in ascii_letters:
            text = str.replace(text, text[i], ' ')
    return text.split()

def make_sentiment(value):
    """Return a sentiment, which represents a value that may not exist.
    """
    assert value is None or (value >= -1 and value <= 1), 'Illegal sentiment value'
    "*** YOUR CODE HERE ***"
    return [value]

def has_sentiment(s):
    """Return whether sentiment s has a value."""
    "*** YOUR CODE HERE ***"
    if s[0] is None:
        return False
    else:
        return True

def sentiment_value(s):
    """Return the value of a sentiment s."""
    assert has_sentiment(s), 'No sentiment value'
    "*** YOUR CODE HERE ***"
    return s[0]

def get_word_sentiment(word):
    """Return a sentiment representing the degree of positive or negative
    feeling in the given word.

    >>> sentiment_value(get_word_sentiment('good'))
    0.875
    >>> sentiment_value(get_word_sentiment('bad'))
    -0.625
    >>> sentiment_value(get_word_sentiment('winning'))
    0.5
    >>> has_sentiment(get_word_sentiment('Berkeley'))
    False
    """
    # Learn more: http://docs.python.org/3/library/stdtypes.html#dict.get
    return make_sentiment(word_sentiments.get(word))

def analyze_tweet_sentiment(tweet):
    """ Return a sentiment representing the degree of positive or negative
    sentiment in the given tweet, averaging over all the words in the tweet
    that have a sentiment value.

    If no words in the tweet have a sentiment value, return
    make_sentiment(None).

    >>> positive = make_tweet('i love my job. #winning', None, 0, 0)
    >>> round(sentiment_value(analyze_tweet_sentiment(positive)), 5)
    0.29167
    >>> negative = make_tweet("saying, 'i hate my job'", None, 0, 0)
    >>> sentiment_value(analyze_tweet_sentiment(negative))
    -0.25
    >>> no_sentiment = make_tweet("berkeley golden bears!", None, 0, 0)
    >>> has_sentiment(analyze_tweet_sentiment(no_sentiment))
    False
    """
    # You may change any of the lines below.
    #average = make_sentiment(None)
    "*** YOUR CODE HERE ***"
    #return average
    words_of_tweet = tweet_words(tweet)
    sent_list = []
    for word in words_of_tweet:
        if has_sentiment(get_word_sentiment(word)):
            sent_list.append(sentiment_value(get_word_sentiment(word)))
        else:
            pass
    if len(sent_list) == 0 or sum(sent_list) == 0:
        return make_sentiment(None)
    else:
        return make_sentiment(sum(sent_list)/len(sent_list))


#################################
# Phase 2: The Geometry of Maps #
#################################

def find_centroid(polygon):
    """Find the centroid of a polygon.

    http://en.wikipedia.org/wiki/Centroid#Centroid_of_polygon

    polygon -- A list of positions, in which the first and last are the same

    Returns: 3 numbers; centroid latitude, centroid longitude, and polygon area

    Hint: If a polygon has 0 area, use the latitude and longitude of its first
    position as its centroid.

    >>> p1, p2, p3 = make_position(1, 2), make_position(3, 4), make_position(5, 0)
    >>> triangle = [p1, p2, p3, p1]  # First vertex is also the last vertex
    >>> round5 = lambda x: round(x, 5) # Rounds floats to 5 digits
    >>> list(map(round5, find_centroid(triangle)))
    [3.0, 2.0, 6.0]
    >>> list(map(round5, find_centroid([p1, p3, p2, p1])))
    [3.0, 2.0, 6.0]
    >>> list(map(float, find_centroid([p1, p2, p1])))  # A zero-area polygon
    [1.0, 2.0, 0.0]
    """
    "*** YOUR CODE HERE ***"
    cx_minus_coefficient = 0
    cy_minus_coefficient = 0
    double_area = 0
    for i in range(len(polygon)):
        if i <= (len(polygon) - 2):
            double_area += (latitude(polygon[i])*longitude(polygon[i+1]) - latitude(polygon[i+1])*longitude(polygon[i]))
    area = double_area * (1/2)
    if area == 0:               #a zero area polygon
        cx = latitude(polygon[0])
        cy = longitude(polygon[0])
    else:
        for i in range(len(polygon)):
            if i <= (len(polygon) - 2):
                cx_minus_coefficient += (latitude(polygon[i]) + latitude(polygon[i + 1])) * (latitude(polygon[i])*longitude(polygon[i+1]) - latitude(polygon[i + 1])*longitude(polygon[i]))
        cx = (1/(6*area)) * cx_minus_coefficient
        for i in range(len(polygon)):
            if i <= (len(polygon) - 2):
                cy_minus_coefficient += (longitude(polygon[i]) + longitude(polygon[i + 1])) * (latitude(polygon[i])*longitude(polygon[i+1]) - latitude(polygon[i + 1])*longitude(polygon[i]))
        cy = (1/(6*area)) * cy_minus_coefficient
    return cx, cy, abs(area)


def find_state_center(polygons):
    """Compute the geographic center of a state, averaged over its polygons.

    The center is the average position of centroids of the polygons in polygons,
    weighted by the area of those polygons.

    Arguments:
    polygons -- a list of polygons

    >>> ca = find_state_center(get_value_from_key(us_states,'CA'))  # California
    >>> round(latitude(ca), 5)
    37.25389
    >>> round(longitude(ca), 5)
    -119.61439

    >>> hi = find_state_center(get_value_from_key(us_states, 'HI'))  # Hawaii
    >>> round(latitude(hi), 5)
    20.1489
    >>> round(longitude(hi), 5)
    -156.21763
    """
    "*** YOUR CODE HERE ***"
    cx_num = 0
    cy_num = 0
    cx_cy_denom = 0
    for i in range(len(polygons)):
        cx_num += list(find_centroid(polygons[i]))[0] * list(find_centroid(polygons[i]))[2]
        cy_num += list(find_centroid(polygons[i]))[1] * list(find_centroid(polygons[i]))[2]
        cx_cy_denom += list(find_centroid(polygons[i]))[2]
    cx = cx_num/cx_cy_denom
    cy = cy_num/cx_cy_denom
    return make_position(cx, cy)


###################################
# Phase 3: The Mood of the Nation #
###################################

def closest_state_to_tweet(tweet):
    #takes in a tweet and returns the two-letter state postal code for the
    #closest state center to that tweet
    database = us_states
    new_dynamic_database = make_database()
    list_of_keys = get_keys(database)
    for i in range(get_len(database)):
        dist = geo_distance(tweet_location(tweet), find_state_center(get_value_from_key(database, list_of_keys[i])))
        new_dynamic_database = add_value(new_dynamic_database, list_of_keys[i], dist)
    distances_from_states = get_values(new_dynamic_database)
    closest_state = ""
    for i in range(get_len(new_dynamic_database)):
        if get_value_from_key(new_dynamic_database, get_keys(new_dynamic_database)[i]) == min(distances_from_states):
            closest_state = get_keys(new_dynamic_database)[i]
    return closest_state

def make_states_database_with_empty_lists():
    database = make_database()
    us_state_key_list = get_keys(us_states)
    for i in range(get_len(us_states)):
        database = add_value(database, us_state_key_list[i], [])
    return database

def group_tweets_by_state(tweets):
    """Return a database that aggregates tweets by their nearest state center.

    The keys of the returned database are state names, and the values are
    lists of tweets that appear closer to that state center than any other.

    tweets -- a sequence of tweet abstract data types

    >>> sf = make_tweet("welcome to san francisco", None, 38, -122)
    >>> ny = make_tweet("welcome to new york", None, 41, -74)
    >>> two_tweets_by_state = group_tweets_by_state([sf, ny])
    >>> get_len(two_tweets_by_state)
    2
    >>> california_tweets = get_value_from_key(two_tweets_by_state, 'CA')
    >>> len(california_tweets)
    1
    >>> tweet_string(california_tweets[0])
    '"welcome to san francisco" @ (38, -122)'
    """
    tweets_by_state = make_states_database_with_empty_lists()
    "*** YOUR CODE HERE ***"
    for i in range(len(tweets)):
        state_of_tweet = closest_state_to_tweet(tweets[i])
        state_list = get_value_from_key(tweets_by_state, state_of_tweet)
        state_list.append(tweets[i])
        tweets_by_state = add_value(tweets_by_state, state_of_tweet, state_list)
    final_tweets_by_state_list = make_database()
    for i in range(get_len(tweets_by_state)):
        if get_values(tweets_by_state)[i] != []:
            final_tweets_by_state_list = add_value(final_tweets_by_state_list, get_keys(tweets_by_state)[i], get_values(tweets_by_state)[i])

    return final_tweets_by_state_list

def average_sentiments(tweets_by_state):
    """Calculate the average sentiment of the states by averaging over all
    the tweets from each state. Return the result as a database from state
    names to average sentiment values (numbers).

    If a state has no tweets with sentiment values, leave it out of the
    database entirely.  Do NOT include states with no tweets, or with tweets
    that have no sentiment, as 0. 0 represents neutral sentiment, not unknown
    sentiment.

    tweets_by_state -- A database from state names to lists of tweets
    """
    averaged_state_sentiments = make_database() 
    "*** YOUR CODE HERE ***"
    for i in range(get_len(tweets_by_state)):
        list_of_tweets_by_state = get_value_from_key(tweets_by_state, get_keys(tweets_by_state)[i])
        list_of_all_sentiments = []
        for j in range(len(list_of_tweets_by_state)):
            if has_sentiment(analyze_tweet_sentiment(list_of_tweets_by_state[j])):
                list_of_all_sentiments.append(sentiment_value(analyze_tweet_sentiment(list_of_tweets_by_state[j])))
        if len(list_of_all_sentiments) > 0:
            average_sentiment_for_state = sum(list_of_all_sentiments)/len(get_value_from_key(tweets_by_state, get_keys(tweets_by_state)[i]))
            averaged_state_sentiments = add_value(averaged_state_sentiments, get_keys(tweets_by_state)[i], average_sentiment_for_state)


    return averaged_state_sentiments


##########################
# Command Line Interface #
##########################

def print_sentiment(text='Are you virtuous or verminous?'):
    """Print the words in text, annotated by their sentiment scores."""
    words = extract_words(text.lower())
    layout = '{0:>' + str(len(max(words, key=len))) + '}: {1:+}'
    for word in words:
        s = get_word_sentiment(word)
        if has_sentiment(s):
            print(layout.format(word, sentiment_value(s)))

def draw_centered_map(center_state='TX', n=10):
    """Draw the n states closest to center_state."""
    us_centers = make_database()
    for state, s in get_items(us_states):
        us_centers = add_value(us_centers, state, find_state_center(s))
    center = get_value_from_key(us_centers, center_state.upper()) 
    dist_from_center = lambda name: geo_distance(center, get_value_from_key(us_centers, name))
    for name in sorted(get_keys(us_centers), key=dist_from_center)[:int(n)]:
        draw_state(get_value_from_key(us_states, name))
        draw_name(name, get_value_from_key(us_centers, name))
    draw_dot(center, 1, 10)  # Mark the center state with a red dot
    wait()

def draw_state_sentiments(state_sentiments):
    """Draw all U.S. states in colors corresponding to their sentiment value.

    Unknown state names are ignored; states without values are colored grey.

    state_sentiments -- A database from state strings to sentiment values
    """
    for name, shapes in get_items(us_states):
        if name in get_keys(state_sentiments):
            sentiment = get_value_from_key(state_sentiments, name)
        else:
            sentiment = None
        draw_state(shapes, sentiment)
    for name, shapes in get_items(us_states):
        center = find_state_center(shapes)
        if center is not None:
            draw_name(name, center)

def draw_map_for_query(term='my job', file_name='tweets2011.txt'):
    """Draw the sentiment map corresponding to the tweets that contain term.

    Some term suggestions:
    New York, Texas, sandwich, my life, justinbieber
    """
    tweets = load_tweets(make_tweet, term, file_name)
    tweets_by_state = group_tweets_by_state(tweets)
    state_sentiments = average_sentiments(tweets_by_state)
    draw_state_sentiments(state_sentiments)
    for tweet in tweets:
        s = analyze_tweet_sentiment(tweet)
        if has_sentiment(s):
            draw_dot(tweet_location(tweet), sentiment_value(s))
    wait()

def swap_tweet_representation(other=[make_tweet_fn, tweet_text_fn,
                                     tweet_time_fn, tweet_location_fn]):
    """Swap to another representation of tweets. Call again to swap back."""
    global make_tweet, tweet_text, tweet_time, tweet_location
    swap_to = tuple(other)
    other[:] = [make_tweet, tweet_text, tweet_time, tweet_location]
    make_tweet, tweet_text, tweet_time, tweet_location = swap_to


@main
def run(*args):
    """Read command-line arguments and calls corresponding functions."""
    import argparse
    parser = argparse.ArgumentParser(description="Run Trends")
    parser.add_argument('--print_sentiment', '-p', action='store_true')
    parser.add_argument('--draw_centered_map', '-d', action='store_true')
    parser.add_argument('--draw_map_for_query', '-m', type=str)
    parser.add_argument('--tweets_file', '-t', type=str, default='tweets2011.txt')
    parser.add_argument('--use_functional_tweets', '-f', action='store_true')
    parser.add_argument('text', metavar='T', type=str, nargs='*',
                        help='Text to process')
    args = parser.parse_args()
    if args.use_functional_tweets:
        swap_tweet_representation()
        print("Now using a functional representation of tweets!")
        args.use_functional_tweets = False
    if args.draw_map_for_query:
        draw_map_for_query(args.draw_map_for_query, args.tweets_file)
        print(args.tweets_file)
        return
    for name, execute in args.__dict__.items():
        if name != 'text' and name != 'tweets_file' and execute:
            globals()[name](' '.join(args.text))
